application/src/main/java/org/openhds/mobile/utilities/OdkCollectHelper.java

```diff
@@ -149,4 +159,101 @@ public static void setStatusComplete(ContentResolver resolver, Uri uri) {
         resolver.update(uri, cv, null, null);
 
     }
+
+    private static String makePlaceholders(int len) {
+        if (len < 1) {
+            // It will lead to an invalid query anyway ..
+            throw new RuntimeException("No placeholders");
+        } else {
+            StringBuilder sb = new StringBuilder(len * 2 - 1);
+            sb.append("?");
+            for (int i = 1; i < len; i++) {
+                sb.append(",?");
+            }
+            return sb.toString();
+        }
+    }
+
+    public static List<FormInstance> getFormInstancesByPath(ContentResolver resolver, Collection<String> ids) {
+
+        ArrayList<FormInstance> formInstances = new ArrayList<FormInstance>();
+
+        Cursor cursor = resolver.query(CONTENT_URI, new String[]{
+                        InstanceProviderAPI.InstanceColumns.INSTANCE_FILE_PATH,
+                        InstanceProviderAPI.InstanceColumns._ID,
+                        InstanceProviderAPI.InstanceColumns.JR_FORM_ID,
+                        InstanceProviderAPI.InstanceColumns.DISPLAY_NAME,
+                        InstanceProviderAPI.InstanceColumns.STATUS},
+                InstanceProviderAPI.InstanceColumns.INSTANCE_FILE_PATH
+                        + " IN (" + makePlaceholders(ids.size()) + ")",
+                ids.toArray(new String[]{}), null);
+
+        if (null == cursor) {
+            return null;
+        }
+
+        while (cursor.moveToNext()) {
+
+            FormInstance formInstance = new FormInstance();
+            String filePath;
+            String formName;
+            String fileName;
+
+            Uri uri;
+
+            filePath = cursor
+                    .getString(cursor
+                            .getColumnIndex(InstanceProviderAPI.InstanceColumns.INSTANCE_FILE_PATH));
+
+            uri = Uri.withAppendedPath(CONTENT_URI, cursor.getString(cursor
+                    .getColumnIndex(InstanceProviderAPI.InstanceColumns._ID)));
+
+            String formStatus = cursor
+                    .getString(cursor
+                            .getColumnIndex(InstanceProviderAPI.InstanceColumns.STATUS));
+
+            formName = cursor
+                    .getString(cursor
+                            .getColumnIndex(InstanceProviderAPI.InstanceColumns.JR_FORM_ID));
+
+            fileName = cursor.getString(cursor.getColumnIndex(InstanceProviderAPI.InstanceColumns.DISPLAY_NAME));
+
+            formInstance.setFilePath(filePath);
+            formInstance.setUriString(uri.toString());
+            formInstance.setFormName(formName);
+            formInstance.setFileName(fileName);
+            formInstances.add(formInstance);
+
+        }
+        cursor.close();
+        return formInstances;
+    }
+
+    public static List<String> getSentFormPaths(ContentResolver resolver, Collection<String> ids) {
```

Comment:
If it helps to understand my earlier comment about using type information and naming: here the naming is more appropriate because it's not entirely clear from type information alone. List&lt;String&gt; isn't clearly a form path.